sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -0,0 +1,623 @@
+    }
+
+    if (update) {
+      val newStats = attrRef.dataType match {
+        case _: NumericType | DateType | TimestampType =>
+          aColStat.copy(distinctCount = newNdv, min = newMin,
+            max = newMax, nullCount = 0)
+        case StringType | BinaryType =>
+          aColStat.copy(distinctCount = newNdv, nullCount = 0)
+      }
+      mutableColStats += (attrRef.exprId -> newStats)
+    }
+
+    // return the filter selectivity.  Without advanced statistics such as histograms,
+    // we have to assume uniform distribution.
+    Some(math.min(1.0, validQuerySet.size / ndv.toDouble))
+  }
+
+  /**
+   * Returns a percentage of rows meeting a binary comparison expression.
+   * This method evaluate expression for Numeric columns only.
+   *
+   * @param op a binary comparison operator uch as =, <, <=, >, >=
+   * @param attrRef an AttributeReference (or a column)
+   * @param literal a literal value (or constant)
+   * @param update a boolean flag to specify if we need to update ColumnStat of a given column
+   *               for subsequent conditions
+   * @return a doube value to show the percentage of rows meeting a given condition
+   */
+  def evaluateBinaryForNumeric(
+      op: BinaryComparison,
+      attrRef: AttributeReference,
+      literal: Literal,
+      update: Boolean)
+    : Option[Double] = {
+
+    var percent = 1.0
+    val aColStat = mutableColStats(attrRef.exprId)
+    val ndv = aColStat.distinctCount
+    val statsRange =
+      Range(aColStat.min, aColStat.max, attrRef.dataType).asInstanceOf[NumericRange]
+
+    val literalValueBD =
+      numericLiteralToBigDecimal(attrRef.dataType, literal.value, literal.dataType)
+
+    // determine the overlapping degree between predicate range and column's range
+    val (noOverlap: Boolean, completeOverlap: Boolean) = op match {
+      case LessThan(l, r) =>
+        (literalValueBD <= statsRange.min, literalValueBD > statsRange.max)
+      case LessThanOrEqual(l, r) =>
+        (literalValueBD < statsRange.min, literalValueBD >= statsRange.max)
+      case GreaterThan(l, r) =>
+        (literalValueBD >= statsRange.max, literalValueBD < statsRange.min)
+      case GreaterThanOrEqual(l, r) =>
+        (literalValueBD > statsRange.max, literalValueBD <= statsRange.min)
+    }
+
+    if (noOverlap) {
+      percent = 0.0
+    } else if (completeOverlap) {
+      percent = 1.0
+    } else {
+      // this is partial overlap case
+      var newMax = aColStat.max
+      var newMin = aColStat.min
+      var newNdv = ndv
+      val literalToDouble = literalValueBD.toDouble
+      val maxToDouble = BigDecimal(statsRange.max).toDouble
+      val minToDouble = BigDecimal(statsRange.min).toDouble
+
+      // Without advanced statistics like histogram, we assume uniform data distribution.
+      // We just prorate the adjusted range over the initial range to compute filter selectivity.
+      // For ease of computation, we convert all relevant numeric values to Double.

```

Comment:
As we use division to compute the percentage, double seems a good balance between speed, space, and ease of computation.